src/main/java/edu/ucla/library/libcal/services/OAuthTokenServiceImpl.java

```diff
@@ -0,0 +1,108 @@
+
+package edu.ucla.library.libcal.services;
+
+import edu.ucla.library.libcal.Config;
+import edu.ucla.library.libcal.Constants;
+import edu.ucla.library.libcal.JsonKeys;
+import edu.ucla.library.libcal.MessageCodes;
+
+import info.freelibrary.util.Logger;
+import info.freelibrary.util.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.auth.User;
+import io.vertx.ext.auth.oauth2.OAuth2Auth;
+
+/**
+ * Implementation of {@link OAuthTokenService}.
+ * <p>
+ * FIXME: the likelihood of {@link OAuthTokenService#getBearerToken(Vertx)} returning an expired token is probably too
+ * high for a production service.
+ */
+public class OAuthTokenServiceImpl implements OAuthTokenService {
+
+    /**
+     * The logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(OAuthTokenServiceImpl.class, MessageCodes.BUNDLE);
+
+    /**
+     * The Vert.x instance.
+     */
+    private final Vertx myVertx;
+
+    /**
+     * The ID of the periodic timer for checking if the OAuth token needs refreshing.
+     */
+    private final long myTimerId;
+
+    /**
+     * The OAuth authentication provider.
+     */
+    private final OAuth2Auth myAuthProvider;
+
+    /**
+     * LibCal credentials represented as an OAuth token.
+     */
+    private User myToken;
+
+    /**
+     * Creates an instance of the service. Use {@link OAuthTokenService#create(Vertx, JsonObject)} to invoke.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aConfig A configuration
+     * @param anAuthProvider An OAuth authentication provider
+     * @param aToken A token that was returned by authenticating against {@code anAuthProvider}
+     */
+    public OAuthTokenServiceImpl(final Vertx aVertx, final JsonObject aConfig, final OAuth2Auth anAuthProvider,
+            final User aToken) {
+        myVertx = aVertx;
+        myAuthProvider = anAuthProvider;
+        myToken = aToken;
+        myTimerId = aVertx.setPeriodic(1000, timerId -> {
```

Comment:
I'm still a little unsure of this polling approach to keeping the token fresh. It seems more reliable to check whether the token has expired before the point it's to be used (and there could be some padding based on timestamp). If it's expired at that point, we could use the secondary token you'd mentioned keeping around and refresh the expired one. That seems more exact and doesn't require the polling. Am I missing an advantage to this approach?

Reply:
More reliable in what sense?

Btw, I don't intend the approach I've taken here to be the final implementation. I chose it because I found it easiest to implement and reason about, and because I think we can improve the implementation without changing the interfaces I'm proposing. I think these interfaces make more sense if one can imagine some future implementation that makes getting an expired token via `OAuthTokenService.getBearerToken` basically impossible (perhaps with token redundancy or some other way), the point being that checking "whether the token has expired before the point it's to be used" is unnecessary due to the guarantees that this future implementation will provide.

Reply:
I went ahead and changed the Jira ticket's title too to reflect that this is the API with a working, but not final, implementation per our discussion earlier today.